pkg/controller/service/service_controller.go

```diff
@@ -582,7 +591,45 @@ func nodeSlicesEqualForLB(x, y []*v1.Node) bool {
 	return nodeNames(x).Equal(nodeNames(x))
 }
 
-func getNodeConditionPredicate() corelisters.NodeConditionPredicate {
+func getNodeTaintPredicate() corelisters.NodePredicate {
+	return func(node *v1.Node) bool {
+		// We add the master to the node list, but its unschedulable.  So we use this to filter
+		// the master.
+		if node.Spec.Unschedulable {
+			return false
+		}
+		// If no taint, accept
+		if len(node.Spec.Taints) == 0 {
+			return true
+		}
+		// As of 1.6, we will taint the master, but not necessarily mark it unschedulable.
+		// Recognize nodes labeled as master, and filter them also, as we were doing previously.
+		if _, hasMasterRoleLabel := node.Labels[LabelNodeRoleMaster]; hasMasterRoleLabel {
+			return false
+		}
+
+		if utilfeature.DefaultFeatureGate.Enabled(kubefeatures.ServiceNodeExclusion) {
+			if _, hasExcludeBalancerLabel := node.Labels[LabelNodeRoleExcludeBalancer]; hasExcludeBalancerLabel {
+				return false
+			}
+		}
+		// Not Ready, no execute taint
+		notReadyTaintTemplate := &v1.Taint{
+			Key:    algorithm.TaintNodeNotReady,
+			Effect: v1.TaintEffectNoExecute,
+		}
+
+		// We consider the node for load balancing only when taint with Effect NoSchedule
+		// or with a specific Taint NotReady, NoExecute
+		if taints.TaintExists(node.Spec.Taints, notReadyTaintTemplate) || taints.TaintExistsWithEffect(node.Spec.Taints, v1.TaintEffectNoSchedule) {
```

Comment:
I believe it's more complex than that. NotReady certainly should be there, but I'd omit NoSchedule part, if we want to be more clever about it, we can do it later.

The reason is that user may apply NoSchedule taint at any time, and thus make a Node dedicated for Pods with appropriate toleration, which may need network connectivity:)

Reply:
I wonder if we have to check if a taint exist, we only searching for effect NoSchedule or NoExecute so we don't check the key, only the Effect

Reply:
We just can't exclude Tainted Nodes.